include/beast/http/concepts.hpp

```diff
@@ -18,63 +21,107 @@ namespace http {
-    using type = decltype(check<T>(0));
-public:
-    static bool constexpr value = type::value;
+    template<class ConstBufferSequence>
+    void
+    operator()(ConstBufferSequence const&);
 };
 
-template<class T, bool B = has_value_type<T>::value>
-struct extract_value_type
-{
-    using type = void;
-};
+template<class T, class = beast::detail::void_t<>>
+struct has_value_type : std::false_type {};
 
 template<class T>
-struct extract_value_type<T, true>
+struct has_value_type<T, beast::detail::void_t<
+    typename T::value_type
+        > > : std::true_type {};
+
+template<class T, class = beast::detail::void_t<>>
+struct has_content_length : std::false_type {};
+
+template<class T>
+struct has_content_length<T, beast::detail::void_t<decltype(
+    std::declval<T>().content_length()
+        )> > : std::true_type
 {
-    using type = typename T::value_type;
+    static_assert(std::is_convertible<
+        decltype(std::declval<T>().content_length()),
+            std::uint64_t>::value,
+        "Writer::content_length requirements not met");
 };
 
-template<class T>
-class has_reader
+#if 0
+template<class T, class M, class = beast::detail::void_t<>>
+struct is_Writer : std::false_type {};
+
+template<class T, class M>
+struct is_Writer<T, M, beast::detail::void_t<decltype(
+    std::declval<T>().init(
+        std::declval<error_code&>())
+    // VFALCO This is unfortunate, we have to provide the template
+    //        argument type because this is not a deduced context?
+    //
+    ,std::declval<T>().template write<detail::write_function>(
+        std::declval<resume_context>(),
+        std::declval<error_code&>(),
+        std::declval<detail::write_function>())
+            )> > : std::integral_constant<bool,
+    std::is_nothrow_constructible<T, M const&>::value &&
+    std::is_convertible<decltype(
+        std::declval<T>().template write<detail::write_function>(
+            std::declval<resume_context>(),
+            std::declval<error_code&>(),
+            std::declval<detail::write_function>())),
+                boost::tribool>::value
+        >
 {
-    template<class U, class R =
-        typename U::reader>
-    static std::true_type check(int);
-    template<class>
-    static std::false_type check(...);
-public:
-    using type = decltype(check<T>(0));
+    static_assert(std::is_same<
+        typename M::body_type::writer, T>::value,
+            "Mismatched writer and message");
 };
 
-template<class T>
-class has_writer
+#else
+
+template<class T, class M>
+class is_Writer
 {
+    template<class U, class R = decltype(
+        std::declval<U>().init(std::declval<error_code&>()),
+            std::true_type{})>
+    static R check1(int);
+    template<class>
+    static std::false_type check1(...);
+    using type1 = decltype(check1<T>(0));
+
+    // VFALCO This is unfortunate, we have to provide the template
+    //        argument type because this is not a deduced context?
+    //
```

Comment:
See boost-dev mailing list.
From "[boost] How can I detect if a template static member function exists?":

_It is not possible AFAIK. The tti library can detect whether a class template is a member of a type but it cannot detect whether a function template is a member of a type. The reason for this, within tti, is that class templates and function templates are not treated orthogonally in C++; a class template can be a template parameter but a function template cannot be a template parameter. If a function template could be a template parameter then tti could detect a function template as a member of a type, whether that function template were a static or a non-static member._

_I have felt that the lack of orthogonality in C++, where a class template can be a template parameter but a function template cannot be a template parameter, is a design weakness of C++. But I would have to convince the C++ standards committee of the practicality of allowing a function template to be a template parameter to possibly change this.  I doubt whether saying that the change would allow the Boost tti library to detect function templates in the same way it can detect class templates, would be a good enough practical reason <g>._